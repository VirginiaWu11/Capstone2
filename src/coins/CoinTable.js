import * as React from "react";
import Paper from "@mui/material/Paper";
import Table from "@mui/material/Table";
import TableBody from "@mui/material/TableBody";
import TableCell from "@mui/material/TableCell";
import TableContainer from "@mui/material/TableContainer";
import TableHead from "@mui/material/TableHead";
import TableRow from "@mui/material/TableRow";
import CardMedia from "@mui/material/CardMedia";
import PushPinIcon from "@mui/icons-material/PushPin";
import PushPinOutlinedIcon from "@mui/icons-material/PushPinOutlined";
import AccountBalanceWalletIcon from "@mui/icons-material/AccountBalanceWallet";
import IconButton from "@mui/material/IconButton";
import Tooltip from "@mui/material/Tooltip";

const columns = [
  {
    id: "image",
    label: "",
    minWidth: 10,
    format: (coin) => (
      <CardMedia
        component="img"
        sx={{
          objectFit: "contain",
          width: "2em",
          display: { sm: "block" },
          mr: 2,
        }}
        image={coin.image}
        alt={`${coin.name} logo`}
      />
    ),
  },
  { id: "name", label: "Name", minWidth: 170 },
  {
    id: "symbol",
    label: "Symbol",
    minWidth: 100,
    format: (value) => value.toUpperCase(),
  },
  {
    id: "current_price",
    label: "Current Price",
    minWidth: 170,
    align: "center",
    format: (value) => "$" + value.toLocaleString("en-US"),
  },
  {
    id: "market_cap",
    label: "Market Cap",
    minWidth: 170,
    align: "center",
    format: (value) => "$" + value.toLocaleString("en-US"),
  },
  {
    id: "total_volume",
    label: "24Hr Volume",
    minWidth: 170,
    align: "center",
    format: (value) => "$" + value.toLocaleString("en-US"),
  },
  {
    id: "fully_diluted_valuation",
    label: "Fully Diluted Valuation",
    minWidth: 170,
    align: "center",
    format: (value) => "$" + value.toLocaleString("en-US"),
  },
  {
    id: "circulating_supply",
    label: "Circulating Supply",
    minWidth: 170,
    align: "center",
    format: (value) => value.toLocaleString("en-US"),
  },
  {
    id: "total_supply",
    label: "Total Supply",
    minWidth: 170,
    align: "center",
    format: (value) => value.toLocaleString("en-US"),
  },
  {
    id: "max_supply",
    label: "Max Supply",
    minWidth: 170,
    align: "center",
    format: (value) => value.toLocaleString("en-US"),
  },
];

const CoinTable = ({
  coins,
  handleCoinModalOpen,
  isPinned,
  handlePin,
  handleUnpin,
  isOnPortfolio,
  removeFromPortfolio,
  handlePortfolioModalOpen,
}) => {
  return (
    <Paper sx={{ width: "100%", overflow: "hidden", margin: 2 }}>
      <TableContainer sx={{ maxHeight: 665 }}>
        <Table stickyHeader aria-label="sticky table">
          <TableHead>
            <TableRow>
              <TableCell></TableCell>
              {columns.map((column) => (
                <TableCell key={column.id} align={column.align}>
                  {column.label}
                </TableCell>
              ))}
            </TableRow>
          </TableHead>
          <TableBody>
            {coins.map((row) => {
              return (
                <TableRow
                  key={row.id}
                  hover
                  role="checkbox"
                  tabIndex={-1}
                  sx={{ cursor: "pointer" }}
                  onClick={() => {
                    handleCoinModalOpen(row);
                  }}
                >
                  {/* <PushPinOutlinedIcon /> */}
                  <TableCell>
                    {isPinned(row.id) ? (
                      <IconButton
                        onMouseDown={(event) => event.stopPropagation()}
                        onClick={(event) => {
                          event.stopPropagation();
                          event.preventDefault();
                          handleUnpin(row.id);
                        }}
                      >
                        <Tooltip title="Unpin from Watchlist">
                          <PushPinIcon color="primary" />
                        </Tooltip>
                      </IconButton>
                    ) : (
                      <IconButton
                        onMouseDown={(event) => event.stopPropagation()}
                        onClick={(event) => {
                          event.stopPropagation();
                          event.preventDefault();
                          handlePin(row.id);
                        }}
                      >
                        <Tooltip title="Pin to Watchlist">
                          <PushPinOutlinedIcon
                            sx={{ transform: "rotate(0.20turn)" }}
                          />
                        </Tooltip>
                      </IconButton>
                    )}
                    {isOnPortfolio(row.id) ? (
                      <IconButton
                        onMouseDown={(event) => event.stopPropagation()}
                        onClick={(event) => {
                          event.stopPropagation();
                          event.preventDefault();
                          removeFromPortfolio(row.id);
                        }}
                      >
                        <Tooltip title="Remove from Portfolio">
                          <AccountBalanceWalletIcon color="primary" />
                        </Tooltip>
                      </IconButton>
                    ) : (
                      <IconButton
                        onMouseDown={(event) => event.stopPropagation()}
                        onClick={(event) => {
                          event.stopPropagation();
                          event.preventDefault();
                          handlePortfolioModalOpen(row);
                        }}
                      >
                        <Tooltip title="Add to Portfolio">
                          <AccountBalanceWalletIcon />
                        </Tooltip>
                      </IconButton>
                    )}
                  </TableCell>
                  {columns.map((column) => {
                    const value = row[column.id];
                    return (
                      <TableCell key={column.id} align={column.align}>
                        {column.format && typeof value === "number"
                          ? column.format(value)
                          : null}
                        {column.format &&
                        typeof value === "string" &&
                        value.match(/^https?:\/\/.+\/.+$/)
                          ? column.format(row)
                          : null}
                        {column.id === "symbol" ? column.format(value) : null}
                        {!column.format ? value : null}
                        {!value ? "N/A" : null}
                      </TableCell>
                    );
                  })}
                </TableRow>
              );
            })}
          </TableBody>
        </Table>
      </TableContainer>
    </Paper>
  );
};

export default CoinTable;
