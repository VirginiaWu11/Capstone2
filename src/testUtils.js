import React from "react";
import { UserContext } from "./auth/UserContext";

const demoUser = {
  username: "testuser",
  first_name: "testfirst",
  last_name: "testlast",
  email: "test@test.net",
};

const UserProvider = ({
  children,
  currentUser = demoUser,
  setCurrentUser,
  token,
  setToken,
  signup,
  signin,
  signout,
  updateProfile,
  infoLoaded,
  setInfoLoaded,
}) => (
  <UserContext.Provider
    value={{
      currentUser,
      setCurrentUser,
      token,
      setToken,
      signup,
      signin,
      signout,
      updateProfile,
      infoLoaded,
      setInfoLoaded,
    }}
  >
    {children}
  </UserContext.Provider>
);

export { UserProvider };
