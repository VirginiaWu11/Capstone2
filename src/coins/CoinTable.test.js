import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router";
import CoinTable from "./CoinTable";
import { UserProvider } from "../testUtils";

it("matches snapshot", function () {
  let coinData = [];
  let coins = [];
  const { asFragment } = render(
    <MemoryRouter>
      <UserProvider>
        <CoinTable coinData={coinData} coins={coins} />
      </UserProvider>
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});
