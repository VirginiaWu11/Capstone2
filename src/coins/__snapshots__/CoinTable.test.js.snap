// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1uxtu1y-MuiPaper-root"
  >
    <div
      class="MuiTableContainer-root css-1qelaid-MuiTableContainer-root"
    >
      <table
        aria-label="sticky table"
        class="MuiTable-root MuiTable-stickyHeader css-xn82ks-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
              scope="col"
            />
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
              scope="col"
            />
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
              scope="col"
            >
              Name
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
              scope="col"
            >
              Symbol
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignCenter MuiTableCell-sizeMedium css-dwuj3p-MuiTableCell-root"
              scope="col"
            >
              Current Price
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignCenter MuiTableCell-sizeMedium css-dwuj3p-MuiTableCell-root"
              scope="col"
            >
              Market Cap
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignCenter MuiTableCell-sizeMedium css-dwuj3p-MuiTableCell-root"
              scope="col"
            >
              24Hr Volume
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignCenter MuiTableCell-sizeMedium css-dwuj3p-MuiTableCell-root"
              scope="col"
            >
              Fully Diluted Valuation
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignCenter MuiTableCell-sizeMedium css-dwuj3p-MuiTableCell-root"
              scope="col"
            >
              Circulating Supply
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignCenter MuiTableCell-sizeMedium css-dwuj3p-MuiTableCell-root"
              scope="col"
            >
              Total Supply
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignCenter MuiTableCell-sizeMedium css-dwuj3p-MuiTableCell-root"
              scope="col"
            >
              Max Supply
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        />
      </table>
    </div>
  </div>
</DocumentFragment>
`;
